File: src/main/java/dev/stocky37/epic7/core/Stats.java
package dev.stocky37.epic7.core;

import com.google.common.collect.ForwardingMap;
import com.google.common.collect.ImmutableMap;
import com.google.common.collect.Streams;

import java.math.BigDecimal;
import java.util.Collection;
import java.util.Map;
import java.util.stream.Collector;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class Stats extends ForwardingMap<Stat, BigDecimal> {
	private final ImmutableMap<Stat, BigDecimal> delegate;

	public Stats() {
		this(ImmutableMap.of());
	}

	public Stats(Map<Stat, BigDecimal> delegate) {
		this(ImmutableMap.copyOf(delegate));
	}

	public Stats(ImmutableMap<Stat, BigDecimal> delegate) {
		this.delegate = delegate;
	}

	public static Stats from(Map<String, BigDecimal> from) {
		return new Stats(from.entrySet().stream()
			.collect(Collectors.toMap(
				entry -> Stat.fromId(entry.getKey()),
				Entry::getValue
			)));
	}

	@Override
	protected Map<Stat, BigDecimal> delegate() {
		return delegate;
	}

	public Stats merge(Map<Stat, BigDecimal> other) {
		return mergeEntries(other.entrySet().stream());
	}

	public Stats merge(Iterable<? extends Map<Stat, BigDecimal>> others) {
		return mergeEntries(flattenMap(others));
	}

	public Stats apply(Map<Stat, BigDecimal> other) throws IllegalStateException {
		ensureCanRunApply();
		return applyEntries(other.entrySet().stream());
	}

	public Stats apply(Iterable<? extends Map<Stat, BigDecimal>> others) throws IllegalStateException {
		ensureCanRunApply();
		return applyEntries(flattenMap(others));
	}

	private void ensureCanRunApply() {
		if(!containsOnlyBaseStats()) {
			throw new IllegalStateException("Can't call apply() on non-base stats");
		}
	}

	private boolean containsOnlyBaseStats() {
		return keySet().stream().allMatch(Stat::isBaseStat);
	}

	private Stats mergeEntries(Stream<Entry<Stat, BigDecimal>> entries) {
		return collectEntries(entries, Collectors.toMap(Entry::getKey, Entry::getValue, BigDecimal::add));
	}

	private Stats applyEntries(Stream<Entry<Stat, BigDecimal>> entries) {
		return collectEntries(entries, Collectors.toMap(entry -> entry.getKey().getBaseStat(), entry -> {
			final Stat stat = entry.getKey();
			final BigDecimal value = entry.getValue();
			return stat.isBaseStat() ? value : getOrDefault(stat.getBaseStat(), BigDecimal.ZERO).multiply(value);
		}, BigDecimal::add));
	}

	@SuppressWarnings("UnstableApiUsage")
	private Stats collectEntries(
		Stream<Map.Entry<Stat, BigDecimal>> entries,
		Collector<Map.Entry<Stat, BigDecimal>, ?, Map<Stat, BigDecimal>> collector
	) {
		return new Stats(Streams.concat(this.entrySet().stream(), entries).collect(collector));
	}

	@SuppressWarnings("UnstableApiUsage")
	private Stream<Map.Entry<Stat, BigDecimal>> flattenMap(Iterable<? extends Map<Stat, BigDecimal>> maps) {
		return Streams.stream(maps).map(Map::entrySet).flatMap(Collection::stream);
	}
}

File: src/main/kotlin/dev/stocky37/epic7/core/Stats.kt
package dev.stocky37.epic7.core

import java.math.BigDecimal

data class Stats(private val stats: Map<Stat, BigDecimal> = mapOf()) : Map<Stat, BigDecimal> by stats {

	companion object {
		fun fromStringMap(stats: Map<String, BigDecimal>): Stats {
			return Stats(stats.mapKeys { Stat.fromId(it.key) })
		}

		fun from(vararg from: Map<Stat, BigDecimal>): Stats {
			return Stats().merge(from.asIterable())
		}

		fun from(from: Iterable<Pair<Stat, BigDecimal>>): Stats {
			return Stats().mergePairs(from.asIterable())
		}
	}

	val isBase: Boolean by lazy {
		stats.keys.all { it.isBaseStat }
	}

	fun merge(from: Iterable<Map<Stat, BigDecimal>>): Stats {
		return mergeEntries(from.map { it.asIterable() }.flatten())
	}

	private fun mergeEntries(from: Iterable<Map.Entry<Stat, BigDecimal>>): Stats {
		return mergePairs(from.map { it.toPair() })
	}

	fun mergePairs(from: Iterable<Pair<Stat, BigDecimal>>): Stats {
		val combined = this.toMutableMap()
		from.forEach { combined.merge(it.first, it.second) { a, b -> a + b } }
		return Stats(combined)
	}

	fun apply(from: Map<Stat, BigDecimal>): Stats {
		if(!isBase) {
			// todo: base stats and regular stats should probs be different classes
			throw IllegalStateException("Can't call apply() on non-base stats")
		}
		val totalStats = this.toMutableMap()
		from.forEach { stat, value ->
			run {
				val newValue = if(stat.isBaseStat) value else stats.getOrDefault(stat.getBaseStat(), BigDecimal.ZERO) * value
				totalStats.merge(stat.getBaseStat(), newValue) { a, b -> a + b }
			}
		}
		return Stats(totalStats)
	}
}
